app/src/main/java/io/homeassistant/companion/android/onboarding/discovery/DiscoveryFragment.kt

```diff
@@ -1,50 +1,119 @@
+            .presenterModule(PresenterModule(this))
+            .build()
+            .inject(this)
+    }
 
     override fun onCreateView(
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View? {
+
+        homeAssistantSearcher = HomeAssistantSearcher(
+            getSystemService(context!!, NsdManager::class.java)!!,
+            this
+        )
+
+        listViewAdapter = object : ArrayAdapter<HomeAssistantInstance>(context!!, R.layout.instance_item, instances) {
+            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+                val v = convertView ?: LayoutInflater.from(context).inflate(
+                    R.layout.instance_item,
+                    null
+                )
+                getItem(position)?.let {
+                    v.findViewById<AppCompatTextView>(R.id.name).text = it.name
+                    v.findViewById<AppCompatTextView>(R.id.url).text = it.url.toString()
+                }
+
+                return v
+            }
+        }
+
         return inflater.inflate(R.layout.fragment_discovery, container, false).apply {
-            viewFlipper = this.findViewById(R.id.view_flipper)
+            findViewById<ListView>(R.id.instance_list_view)?.apply {
+                adapter = listViewAdapter
+                setOnItemClickListener { _, _, position, _ -> presenter.onUrlSelected(instances[position].url) }
+            }
+
             this.findViewById<Button>(R.id.manual_setup)
                 .setOnClickListener { (activity as DiscoveryListener).onSelectManualSetup() }
-            this.findViewById<Button>(R.id.retry).setOnClickListener { scan() }
         }
     }
 
-    override fun onStart() {
-        super.onStart()
-        scan()
+    override fun onResume() {
+        super.onResume()
+        homeAssistantSearcher.beginSearch()
+    }
+
+    override fun onPause() {
+        super.onPause()
+        homeAssistantSearcher.stopSearch()
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        homeAssistantSearcher.stopSearch()
+        presenter.onFinish()
+    }
+
+    override fun onUrlSaved() {
+        (activity?.application as GraphComponentAccessor).urlUpdated()
+        (activity as DiscoveryListener).onHomeAssistantDiscover()
+    }
+
+    override fun onInstanceFound(instance: HomeAssistantInstance) {
+        if (!instances.contains(instance)) {
+            instances.add(instance)
+            activity?.runOnUiThread {
+                listViewAdapter.notifyDataSetChanged()
+            }
+        }
     }
 
-    private fun scan() {
-        viewFlipper.displayedChild = LOADING_VIEW
-        Handler().postDelayed({
-            viewFlipper.displayedChild = TIMEOUT_VIEW
-        }, 5000)
+    override fun onScanError() {
+        Toast.makeText(context, R.string.failed_scan, Toast.LENGTH_LONG).show()
```

Comment:
When a scan error occurs, we should tell the user, and then show the "manually enter address" view.

Reply:
Updated so that if the scan failed and we haven't found any instances yet to go to the manual entry, otherwise we should have an instance in the list to select.